### PR TITLE
Round M2 dot product in Weyl decomposition

### PR DESCRIPTION
**Pull request by mtreinish**
### Summary

Differences in floating point precision between different systems and
environments was causing inconsistent decompositions for decompositions
with 0s in the unitary would have differing results because of very
subtle floating point precision errors in the results. For example one
system would return 0 for a value and another would return
-1.55582133e-19 (which is essentially 0). This would have compound
effects in later stages because the signs were different. This commit
attempts to fix cases like this by rounding to 13 decimal places to
prevent issues like this in the future.

### Details and comments

### CODE REVIEW

File: qiskit/quantum_info/synthesis/two_qubit_decompose.py [hunk 1]
```diff
@@ -161,7 +161,7 @@ def __init__(self, unitary_matrix):
         U *= la.det(U)**(-0.25)
 
         Up = _Bd.dot(U).dot(_B)
-        M2 = Up.T.dot(Up)
+        M2 = np.around(Up.T.dot(Up), 13)
```
Comment (nonhermitian) [inline qiskit/quantum_info/synthesis/two_qubit_decompose.py:164]: I see we use `np.around` a lot now in the code but it is a bit non-standard.  Would be nice to just do something like:

```python

if abs(M2) < 1e-15:
    M2 = 0
```
where `1e-15` can be replaced with any other `eps` value of ones choice.

Comment (mtreinish) [reply qiskit/quantum_info/synthesis/two_qubit_decompose.py]: Done in: https://github.com/Qiskit/qiskit-terra/pull/4835/commits/a6ab027af035b3cee40fedddda39745456e5cac8

**Comment by levbishop** [comment]
In general I'm loth to add these kind of roundings because it usually indicates an underlying problem being papered over. At the very least when I come across code with round-to-epsilons in it, that indicates to me scars from previous battles with numeric bugs and makes me very reluctant to make changes in order to avoid reopening old wounds. To reduce such fears can we at least capture the problematic cases as a test instance and add a comment with a pointer at the lines with the rounding?

I've been playing around off and on with a couple other implementations for Weyl decomposition to eliminate the ugly randomization anyway, so with a test case I'll be able to check if any of them is able to solve address this issue without rounding.

**Comment by mtreinish** [comment]
> In general I'm loth to add these kind of roundings because it usually indicates an underlying problem being papered over. At the very least when I come across code with round-to-epsilons in it, that indicates to me scars from previous battles with numeric bugs and makes me very reluctant to make changes in order to avoid reopening old wounds. To reduce such fears can we at least capture the problematic cases as a test instance and add a comment with a pointer at the lines with the rounding?
> 
> I've been playing around off and on with a couple other implementations for Weyl decomposition to eliminate the ugly randomization anyway, so with a test case I'll be able to check if any of them is able to solve address this issue without rounding.

Sure I'll add a test case for the example that prompted me to add this. I definitely know that pain of debugging numeric bugs. I spent way too much time digging through things here to figure out that it was the CX gate decomposition in the basis being subtly off of zero here for some envs that was causing different decompositions on different systems and not an issue with the random unitary being tested..

**Comment by nonhermitian** [comment]
Not sure what the concern is here. Setting was is a floating point zero to exactly zero is fairly common. Especially if the sign is important later.

**Comment by levbishop** [comment]
Agree that it's common. IMO it should be much less common and used as a last resort.
It increases the error of the routine (by in this case factor of around 1E6. If the results of that routine feed into other routines with similar epsilon tolerances then those will have to broaden their tolerances. In the worst case you feed the output of this routine into a downstream routine that also needs to round-to-epsilon and because of this needs to broaden its tolerances correspondingly.
A factor of 1E6 here, a factor of 1E6 there, pretty soon you're talking real money.

**Pull request merged**